### PR TITLE
KT-7710 Intention to convert lambda to anonymous function

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-7710

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/intentions/LambdaToAnonymousFunctionIntention.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        if (returnType.isNotEmpty()) {
+            val lastStatement = bodyExpression.statements.lastOrNull()
+            if (lastStatement != null && lastStatement !is KtReturnExpression) {
+                val foldableReturns = BranchedFoldingUtils.getFoldableReturns(lastStatement)
+                if (foldableReturns == null || foldableReturns.isEmpty()) {
+                    lastStatement.replace(psiFactory.createExpressionByPattern("return $0", lastStatement))
+                }
+            }
+        }
+        val anonymousFunction = element.replaced(
```
[the remark this answers is not in the document]
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/LambdaToAnonymousFunctionIntention.kt]: `CallableBuilder` will break format:

idea/testData/intentions/lambdaToAnonymousFunction/argument.kt
```
// expected
fun test() {
    foo(fun(it: Int): String { return "" })
}
// actual
fun test() {
    foo(fun(it: Int): String {
        return ""
    })
}
```

idea/testData/intentions/lambdaToAnonymousFunction/hasComment.kt
```
// expected
fun test() {
    foo(fun(it: Int): String {
        // comment1
        return ""
        // comment2
    })
}
// actual
fun test() {
    foo(fun(it: Int): String {
// comment1
        return ""
        // comment2
    })
}
```

File: idea/src/org/jetbrains/kotlin/idea/intentions/LambdaToAnonymousFunctionIntention.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    "Convert to anonymous function",
+    "Convert lambda expression to anonymous function"
+) {
+
+    override fun isApplicableTo(element: KtLambdaExpression, caretOffset: Int): Boolean {
+        val argument = element.getParentOfTypesAndPredicate(true, KtValueArgumentList::class.java, KtLambdaArgument::class.java) { true }
+                ?: return false
+
+        val callElement = argument.parent as? KtCallElement ?: return false
+        if (callElement.getCalleeExpressionIfAny() !is KtNameReferenceExpression) return false
```
[the remark this answers is not in the document]
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/LambdaToAnonymousFunctionIntention.kt]: Deleted this check.

File: idea/src/org/jetbrains/kotlin/idea/intentions/LambdaToAnonymousFunctionIntention.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return descriptor.valueParameters.none { it.name.isSpecial }
+    }
+
+    override fun applyTo(element: KtLambdaExpression, editor: Editor?) {
+        val functionLiteral = element.functionLiteral
+        val bodyExpression = functionLiteral.bodyExpression ?: return
+        val descriptor = functionLiteral.descriptor as? AnonymousFunctionDescriptor ?: return
+        val psiFactory = KtPsiFactory(element)
+
+        bodyExpression.collectDescendantsOfType<KtReturnExpression>().forEach { it.labeledExpression?.delete() }
```
[the remark this answers is not in the document]
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/LambdaToAnonymousFunctionIntention.kt]: Fixed

File: idea/src/org/jetbrains/kotlin/idea/intentions/LambdaToAnonymousFunctionIntention.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import org.jetbrains.kotlin.types.typeUtil.isUnit
+
+class LambdaToAnonymousFunctionIntention : SelfTargetingIntention<KtLambdaExpression>(
+    KtLambdaExpression::class.java,
+    "Convert to anonymous function",
+    "Convert lambda expression to anonymous function"
+) {
+
+    override fun isApplicableTo(element: KtLambdaExpression, caretOffset: Int): Boolean {
+        val argument = element.getParentOfTypesAndPredicate(true, KtValueArgumentList::class.java, KtLambdaArgument::class.java) { true }
```
[the remark this answers is not in the document]
Comment (t-kameyama) [reply idea/src/org/jetbrains/kotlin/idea/intentions/LambdaToAnonymousFunctionIntention.kt]: Fixed

**Comment by mglukhikh** [comment]
> CallableBuilder will break format
The situation you described is a formatter bug. See https://youtrack.jetbrains.com/issue/KT-24360

**Comment by mglukhikh** [comment]
Finally merged (with lot of small fixes) into 1.2.60. Thank you.

**Pull request closed without merging**